tests/test_streaming.py

```diff
@@ -59,6 +59,81 @@ def test_dataloader_epoch_size_no_streams(local_remote_dir: Tuple[str,
             assert samples_seen == epoch_size
 
 
+# @pytest.mark.parametrize('drop_last', [False, True])
+# @pytest.mark.parametrize('num_workers', [3, 6])
+# @pytest.mark.parametrize('num_canonical_nodes', [4, 8])
+
+
+@pytest.mark.parametrize('batch_size', [4])
+@pytest.mark.parametrize('seed', [2222])
+@pytest.mark.parametrize('shuffle', [True])
+@pytest.mark.parametrize('drop_last', [True])
+@pytest.mark.parametrize('num_workers', [1])
+@pytest.mark.parametrize('num_canonical_nodes', [4])
+@pytest.mark.parametrize('num_stream_1_samples', [200, 255])
+@pytest.mark.parametrize('num_stream_2_samples', [342, 557])
+@pytest.mark.usefixtures('local_remote_dir')
+def test_dataloader_stratified_batching(local_remote_dir: Tuple[str, str], batch_size: int,
+                                        seed: int, shuffle: bool, drop_last: bool,
+                                        num_workers: int, num_canonical_nodes: int,
+                                        num_stream_1_samples: int, num_stream_2_samples: int):
+    # create mock datasets for 2 streams. Second one has 1.5x the samples
+    local, remote = local_remote_dir
+    local1 = os.path.join(local, 'stream1')
+    local2 = os.path.join(local, 'stream2')
+    remote1 = os.path.join(remote, 'stream1')
+    remote2 = os.path.join(remote, 'stream2')
+
+    # stream 1 has samples 0->num_stream_1_samples*3
+    convert_to_mds(out_root=remote1,
+                   dataset_name='sequencedataset',
+                   num_samples=num_stream_1_samples,
+                   size_limit=1 << 8)
+    # stream 2 has samples num_stream_1_samples*3 and above. This lets us differentiate between the samples from each stream
+    convert_to_mds(out_root=remote2,
+                   dataset_name='sequencedataset',
+                   num_samples=num_stream_2_samples,
+                   offset=num_stream_1_samples * 3,
+                   size_limit=1 << 8)
+
+    stream1 = Stream(local=local1, remote=remote1)
```

Comment:
Can you also add a test case where user controls the number of samples per stream in a batch ?

Reply:
done!